**Baseline Model**

In this notebook we are going to build a reccomendation system only based on the "genres" without the including the NLP techniques on "description"

In [33]:
import sys
import os

project_root = os.path.abspath(os.path.join(os.getcwd(), '..'))

if project_root not in sys.path:
    sys.path.append(project_root)

from src import config
WEBTOON_DATA_PATH = config.WEBTOON_DATA_PATH

In [34]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

In [35]:
try:
    df = pd.read_csv(WEBTOON_DATA_PATH)

except FileNotFoundError:
    print("File not found. Please check the file path and try again.")

In [36]:
df_baseline = df[['title_romaji' , 'genres']].copy()
print(f"Data shape: {df_baseline.shape}")
df_baseline.head()

Data shape: (18250, 2)


,title_romaji,genres
0,Chainsaw Man,"['Action', 'Comedy', 'Drama', 'Horror', 'Super..."
1,Na Honjaman Level Up,"['Action', 'Adventure', 'Fantasy']"
2,Jujutsu Kaisen,"['Action', 'Drama', 'Supernatural']"
3,Berserk,"['Action', 'Adventure', 'Drama', 'Fantasy', 'H..."
4,Shingeki no Kyojin,"['Action', 'Drama', 'Fantasy', 'Mystery']"


**Basic Modelling (CountVectorizer)**

-Only look for the interested tag not the importance of the tag



**Why is is suitable for baseline**

-Fast and the computing cost is low

In [37]:
count_vectorizer = CountVectorizer(stop_words='english')
count_matrix = count_vectorizer.fit_transform(df_baseline['genres'])
print(f"Count Matrix shape: {count_matrix.shape}")

Count Matrix shape: (18250, 22)


In [38]:
cosine_sim_baseline = cosine_similarity(count_matrix, count_matrix)

In [39]:
indices = pd.Series(df_baseline.index, index=df_baseline['title_romaji']).drop_duplicates()

In [40]:
def get_baseline_rec(title, cosine_sim = cosine_sim_baseline):
    if title not in indices:
        return "Title not found in the dataset."
    idx = indices[title]

    if isinstance(idx, pd.Series):
        idx = idx.iloc[0]

    sim_scores = list(enumerate(cosine_sim[idx]))

    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    sim_scores = sim_scores[1:6]

    rec_indices = [i[0] for i in sim_scores]

    return df_baseline['title_romaji'].iloc[rec_indices]

**Baseline Outcomes**

In [41]:
test_title = df_baseline['title_romaji'].iloc[10]

In [42]:
print(f"Chosen Title: {test_title}\n")
print("Top 5 Recommendations:")
print(get_baseline_rec(test_title))

Chosen Title: One Punch-Man

Top 5 Recommendations:
95                Ansatsu Kyoushitsu
255                    Undead Unluck
729                    One Punch Man
5089    Xianwang De Richang Shenghuo
8457                    Ankoku Delta
Name: title_romaji, dtype: object


In [52]:
def calculate_genre_match_score(input_title, df_subset):
    input_tags = df_subset[df_subset['title_romaji'] == input_title]['genres'].iloc[0]
    input_tags_set = set([t.strip().lower() for t in str(input_tags).split(',')])

    match_scores = []

    for _, row in df_subset.iterrows():
        rec_tags_set = set([t.strip().lower() for t in str(row['genres']).split(',')])

        intersection = input_tags_set.intersection(rec_tags_set)
        score = len(intersection) / len(input_tags_set) if len(input_tags_set) > 0 else 0
        match_scores.append(score)

    return np.mean(match_scores)


In [53]:
total_scores = []
random_titles = df_baseline['title_romaji'].sample(100).tolist()

In [55]:
for title in random_titles:
    score = calculate_genre_match_score(title, df_baseline)
    total_scores.append((title, score))

In [57]:
scores_only = [score for _, score in total_scores]
average_genre_match = np.mean(scores_only)
print(f"Average Genre Match: %{average_genre_match*100:.2f}")

Average Genre Match: %14.45


**Outcome Evaluation**

Baseline model is only interested in genre similartiy so the reccomendations may be irrevelent to the chosen title's content. It also produces a n Average Genre Match  of only %14.45 which means most of the reccomended titles share very few genre tags with the input title

**Shortcomings**

-Description is missing

-Popularity, format not included

-Sentiments are not included